### ISSUE TITLE
diff algorithm

### ISSUE THREAD

**Issue by telephon** [supercollider/supercollider#3204]
It would be very useful to have a diff algorithm in the common class library.

It would solve problems like #3020, and would allow implementations like Steno https://github.com/telephon/Steno/blob/master/Classes/DiffString.sc to be independent from system commands.

This is a solved problem and well tested in versioning systems like git. The most common solution is the Myers algorithm. There is a Ruby implementation.

These blog entries explain it:
https://blog.jcoglan.com/2017/02/12/the-myers-diff-algorithm-part-1/
https://blog.jcoglan.com/2017/02/15/the-myers-diff-algorithm-part-2/
https://blog.jcoglan.com/2017/02/17/the-myers-diff-algorithm-part-3/

It might be good to make it a primitive, but perhaps we can take advantage of the flexibility of an sclang implementation.

**Comment by cianoc** [comment]
Git uses libxdiff. Why not build a primitive around that?

**Comment by jamshark70** [comment]
For #3020, would it be more efficient to concatenate all the keys into a CR delimited string and use an existing text-diff library, or to write something that can work directly on SC arrays?

**Comment by telephon** [comment]
yes, in Steno I use text diff  via command line. Is ok for many purposes, better at least to have it as a primitive, but an sclang version allows us to use equality to match.

**Comment by telephon** [comment]
But the Ruby version of the Myers algorithm is almost literally sclang, just the syntax is different. So perhaps it is worth a try to port it from that?

**Comment by scztt** [comment]
If the work is being done to incorporate a diff library, why not just do libgit2 instead? It provides diffing - but we'd also have the potential to access other git functionality. This would be helpful for quarks, and could.be used to add some git-intelligence to the ide.
